### PR TITLE
Package non-rust objects

### LINKED ISSUE #65335: 3rd party objects missing from nightly-2019-10-12
```
$ rustup toolchain add nightly-2019-10-11
$ rustup toolchain add nightly-2019-10-12
$ rustup target add --toolchain nightly-2019-10-11 x86_64-fortanix-unknown-sgx
$ rustup target add --toolchain nightly-2019-10-12 x86_64-fortanix-unknown-sgx
$ ls .rustup/toolchains/nightly-2019-10-11-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-fortanix-unknown-sgx/lib/libunwind.a
.rustup/toolchains/nightly-2019-10-11-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-fortanix-unknown-sgx/lib/libunwind.a
$ ls .rustup/toolchains/nightly-2019-10-12-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-fortanix-unknown-sgx/lib/libunwind.a
ls: cannot access '.rustup/toolchains/nightly-2019-10-12-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-fortanix-unknown-sgx/lib/libunwind.a': No such file or directory
```

### PR DESCRIPTION
**Pull request by jethrogb**
Fixes #65335

I'd like to use this opportunity to talk about increasing test coverage of tier 2 targets to at least make sure a "hello world" binary can be succesfully linked. There have been similar issues for individual targets in the past, but now we have one affecting several tier 2 targets (and even a tier 1 target).

r? @Mark-Simulacrum
cc @cuviper

### CODE REVIEW

**Comment by jethrogb** [comment]
Not sure how to properly test this.

**Comment by pietroalbini** [comment]
The best way to test this I think is to open another PR with the same commits as this one but targeting a `tmp-publish` branch instead of `master`. Then we can r+ it, and once it's merged we can publish the build on dev-static so it's installable with rustup for all the platforms we distribute.

This is going to take time though, so I'd like (as proposed in https://github.com/rust-lang/rust/issues/65335#issuecomment-541328516) to rollback https://github.com/rust-lang/rust/pull/64823 in the meantime.

**Comment by Mark-Simulacrum** [comment]
The changes here look reasonable, even if it is presumably a bit unknown as to whether they actually fully fix things or not. I'm going to @bors r+ p=1 this PR with the goal of landing it nightly tonight. I believe we have enough time left that we can do the following:

 * land this PR, get a dev-static release with it, check if that fixes things -- if so, stop here
 * if things are not fixed by this PR, land a revert to it and the rustc-dev component PR, fixing tonight's nightly

This seems like a fast and reasonable approach to getting a working nightly tonight no matter what.

**Comment by jethrogb** [comment]
I've seen this error in my internal CI when the source file inside `copy_third_party_objects` doesn't exist.

File: src/bootstrap/lib.rs [hunk 1]
```diff
@@ -1144,6 +1144,7 @@ impl Build {
     pub fn copy(&self, src: &Path, dst: &Path) {
         if self.config.dry_run { return; }
         self.verbose_than(1, &format!("Copy {:?} to {:?}", src, dst));
+        if src == dst { return; }
```
Comment (jethrogb) [inline src/bootstrap/lib.rs]: Not sure why this is now necessary, but hey, it appears to work!

**Comment by gnzlbg** [comment]
> I'd like to use this opportunity to talk about increasing test coverage of tier 2 targets to at least make sure a "hello world" binary can be succesfully linked. There have been similar issues for individual targets in the past, but now we have one affecting several tier 2 targets (and even a tier 1 target).

+1. 

Any chance of merging this soon ?

**Comment by jethrogb** [comment]
@gnzlbg the PR that broke everything was reverted in #65342, tonight's nightly should work again. I don't think people are in a hurry to re-land it.

**Comment by Mark-Simulacrum** [comment]
Closing in favor of https://github.com/rust-lang/rust/pull/65474 which includes this.

**Pull request closed without merging**